File: problems/python/src/palindrome_number.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_palindrome(self, x: int) -> bool:
        pass


class ReverseString(Problem):
    @override
    def is_palindrome(self, x: int) -> bool:
        return str(x) == str(x)[::-1]


class CompareHalfString(Problem):
    @override
    def is_palindrome(self, x: int) -> bool:
        s = str(x)
        start = 0
        end = len(s) - 1
        while start < end:
            if s[start] != s[end]:
                return False
            start += 1
            end -= 1
        return True


SOLUTIONS: list[Problem] = [ReverseString(), CompareHalfString()]

File: problems/src/main/java/PalindromeNumber.java
public enum PalindromeNumber {
    REVERSE_STRING {
        @Override
        public boolean isPalindrome(int x) {
            String s = String.valueOf(x);
            return s.contentEquals(new StringBuilder(s).reverse());
        }
    },
    COMPARE_HALF_STRING {
        @Override
        public boolean isPalindrome(int x) {
            String s = String.valueOf(x);
            int start = 0;
            int end = s.length() - 1;
            while (start < end) {
                if (s.charAt(start++) != s.charAt(end--)) {
                    return false;
                }
            }
            return true;
        }
    };

    public abstract boolean isPalindrome(int x);
}
